### PR TITLE
[luv-277] feat: add Gemini CLI integration (beta)

### PR DESCRIPTION
**Pull request by NiveditJain**
## Summary

Adds **Gemini CLI** as the seventh supported agent CLI, joining Claude Code, OpenAI Codex, GitHub Copilot, Cursor Agent, OpenCode, and Pi. Verified end-to-end against `gemini-cli` v0.40.1.

- `--cli gemini` writes Claude-shape hook entries into `~/.gemini/settings.json` (user) or `<cwd>/.gemini/settings.json` (project) using Gemini's `{matcher, hooks: [{type, command, timeout}]}` matcher-wrapper schema.
- All **11 documented Gemini events** install: `SessionStart`, `SessionEnd`, `BeforeAgent`, `AfterAgent`, `BeforeModel`, `AfterModel`, `BeforeToolSelection`, `BeforeTool`, `AfterTool`, `PreCompress`, `Notification`. Three of those (`BeforeModel` / `AfterModel` / `BeforeToolSelection`) lack a Claude canonical equivalent — the binary still records activity for them so future policies can opt in.
- **Tool-name canonicalization** via `GEMINI_TOOL_MAP` translates Gemini's snake_case names (`run_shell_command`, `read_file`, `read_many_files`, `write_file`, `replace`, `glob`, `grep_search`, `list_directory`, `web_fetch`, `google_web_search`, `write_todos`, `save_memory`, `ask_user`) to Claude PascalCase (`Bash`, `Read`, `Write`, `Edit`, `Glob`, `Grep`, `LS`, `WebFetch`, `WebSearch`, `TodoWrite`, `Memory`, `AskUser`) so existing builtin policies (block-sudo, block-rm-rf, sanitize-api-keys, …) fire unchanged on Gemini sessions. MCP tool names (`mcp_<server>_<tool>`) and Skills tool names pass through unchanged.
- **Response-shape adapter** in `policy-evaluator.ts` emits Gemini's flat `{decision: "deny", reason}` deny shape (preferred per Gemini's "Golden Rule" exit-0 contract), `{hookSpecificOutput: {hookEventName, additionalContext}}` for context injection on `BeforeAgent` / `AfterTool` / `SessionStart`, and `{decision: "block", reason}` on `AfterAgent` for force-retry semantics matching Claude's exit-2-from-Stop "do this before stopping" pattern. The agent-emitted `hookEventName` round-trips through a new `SessionMetadata.rawHookEventName` field captured before canonicalization, and the deny-message `displayTool` is event-aware so non-tool events emit "Blocked prompt"/"Blocked session start"/etc. instead of the misleading "Blocked unknown tool".
- **Path-protection** (`isAgentInternalPath` + `isAgentSettingsFile`) covers `~/.gemini/` and `.gemini/settings.json`.
- **Frontend / dashboard parity**: `lib/cli-registry.ts` adds a `Gemini CLI` entry with a sky-blue badge; `lib/projects.ts` merges Gemini projects into `/projects` via new `lib/gemini-projects.ts` (scans `~/.gemini/tmp/<basename>/chats/session-<ts>-<uuid-prefix>.jsonl`, recovers cwd from sibling `.project_root` text marker, names rows by `encodeFolderName(cwd)` so cross-CLI merge unions on the same key); new `lib/gemini-sessions.ts` parses the JSONL transcript schema with a permissive timestamp regex and a bounded 4 KB `readFirstLineSync` for header-only validation; `app/project/[name]/page.tsx`, `app/project/[name]/session/[sessionId]/page.tsx`, and `app/policies/hooks-client.tsx` extend their per-CLI fallback chains.
- **Telemetry**: adding `"gemini"` to `IntegrationType` automatically tags all six existing PostHog events (`hooks_installed`, `hooks_removed`, `hook_policy_triggered`, `custom_hook_error`, `custom_hooks_loaded`, `convention_policies_loaded`) with `cli: "gemini"`. Activity-store entries get `integration: "gemini"`. Handler tests prove canonicalized tool names (e.g. `"Bash"`, not raw `"run_shell_command"`) reach telemetry.
- **Dogfood**: ships this repo's own `.gemini/settings.json` so contributors using `gemini` get hooks active automatically. Uses `$GEMINI_PROJECT_DIR` for resolver stability (Gemini also sets `$CLAUDE_PROJECT_DIR` as a back-compat alias).
- **README** restructured into two centred logo rows + isolated "+ more coming soon" line so the seventh logo doesn't crowd the layout. Light/dark Gemini SVGs at `assets/logos/gemini-{light,dark}.svg`.

## End-to-end live verification

Verified against `gemini-cli` v0.40.1 by dropping `.gemini/settings.json` into this repo and running `gemini --skip-trust --yolo -p "say hi"`. Confirmed:

1. Gemini picked up the dogfood `.gemini/settings.json` and called the failproofai binary on every event including `BeforeAgent`, `AfterAgent`, and `SessionEnd`.
2. The `bun $GEMINI_PROJECT_DIR/bin/failproofai.mjs --hook ... --cli gemini` command form works as expected.
3. The `require-commit-before-stop` builtin policy fired on `AfterAgent`, returning `{"decision": "block", "reason": "MANDATORY ACTION REQUIRED ..."}`.
4. Gemini interpreted that `block` decision and **actually retried** — attempting to commit before stopping, exactly as the Stop-policy semantics require.

## Test plan

- [x] `bun run test:run` — **1432/1432 unit tests pass** (was 1424; CR follow-up added 5 evaluator regression tests + 3 projects-aggregation tests), including new Gemini-specific cases for: install/uninstall round-trip, idempotency, marker preservation, scope-path resolution, event-name canonicalization (all 11 events), tool-name canonicalization (every entry in GEMINI_TOOL_MAP + unknown-name passthrough), deny / instruct / allow-with-context shapes, raw-hookEventName fallback, event-appropriate deny messages, telemetry tagging with `cli: "gemini"`, activity-store `integration: "gemini"` round-trip, and Gemini aggregation (Gemini-only inclusion, cross-CLI merge by encoded slug, reject-fallback).
- [x] `bun run test:e2e` — **285/285 e2e tests pass** (4 pre-existing skipped, gated by piPresent / opencode availability), including a new `__tests__/e2e/hooks/gemini-integration.e2e.test.ts` with 18 cases covering the four Gemini-distinguishing invariants (tool-name canonicalization, event-name canonicalization, response-shape correctness, edge cases like 1MB+ stdin and malformed JSON).
- [x] `bun run lint` — clean (1 pre-existing warning unrelated to Gemini).
- [x] `bunx tsc --noEmit` — clean.
- [x] `bun run build` — clean (Next.js + dist bundle).
- [x] **Live `gemini` v0.40.1 integration test** — dogfood `.gemini/settings.json` picked up; `require-commit-before-stop` fired and Gemini retried.
- [x] Regression coverage — every existing CLI's tests (Claude / Codex / Copilot / Cursor / OpenCode / Pi) still passes unchanged, including settings-file shape snapshots.
- [x] **CodeRabbit review feedback addressed** — all 6 inline comments fixed in commit a977cd9 with reply on each thread (CR1 raw hookEventName + event-specific deny message, CR2 Gemini aggregation tests, CR3 PR-suffix in CHANGELOG, CR4 critical project name encoding, CR5 looser SESSION_FILE_RE, CR6 bounded first-line read).

## Files

Three commits:
- `223feb0` — foundation: types, integrations, handler, policy-evaluator, builtin-policies, install-prompt, resolve-permission-mode, bin/failproofai.mjs, dogfood `.gemini/settings.json`, in-place test extensions.
- `2e43d26` — polish: README two-row layout, CHANGELOG, CLAUDE.md section + capability matrix, docs/*.mdx, dashboard wiring under `lib/` and `app/`, light/dark Gemini SVGs, new `__tests__/e2e/hooks/gemini-integration.e2e.test.ts`.
- `a977cd9` — CodeRabbit review follow-ups: raw hookEventName fallback + event-aware deny message, encodeFolderName for project name, looser session filename regex, bounded header read, three Gemini aggregation tests, CHANGELOG PR-suffix.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/hooks/policy-evaluator.ts [hunk 1]
```diff
@@ -317,6 +344,72 @@ export async function evaluatePolicies(
...
+      // Map back from canonical → Gemini event name. Prefer the raw event name
+      // off the session (handler.ts populates it from parsed.hook_event_name)
+      // so we don't have to maintain a reverse lookup table.
+      const supportsContext =
+        eventType === "UserPromptSubmit" ||
+        eventType === "PreToolUse" ||
+        eventType === "PostToolUse" ||
+        eventType === "SessionStart";
+      if (supportsContext) {
+        const hookEventName = session?.hookEventName ?? eventType;
```
[the remark this answers is not in the document]
Comment (NiveditJain) [reply src/hooks/policy-evaluator.ts]: Fixed in a977cd9 (`policy-evaluator.ts` + `handler.ts` + `types.ts`):

- Added `SessionMetadata.rawHookEventName`, captured in `handler.ts` from the raw `--hook` CLI arg before canonicalization. The Gemini context-injection branches in `policy-evaluator.ts` now use `session?.hookEventName ?? session?.rawHookEventName ?? eventType` so the response's `hookEventName` round-trips correctly even when stdin omits `hook_event_name`.
- Branched `displayTool` in the deny path on `eventType` so non-tool events emit event-appropriate copy: `Blocked prompt` (UserPromptSubmit), `Blocked session start` (SessionStart), `Blocked session end` (SessionEnd), `Blocked stop` (Stop), `Blocked operation` (otherwise) — instead of the misleading `Blocked unknown tool`.
- Added 5 regression tests in `__tests__/hooks/policy-evaluator.test.ts`: rawHookEventName fallback for instruct, same fallback for allow-with-context, hookEventName-from-stdin still wins when both set, UserPromptSubmit deny says "Blocked prompt", SessionStart deny says "Blocked session start". 1432/1432 unit + 285 e2e green.

File: __tests__/lib/projects.test.ts [hunk 2]
```diff
@@ -40,19 +40,25 @@ vi.mock("@/lib/pi-projects", () => ({
   getPiProjects: vi.fn(async () => []),
 }));
 
+vi.mock("@/lib/gemini-projects", () => ({
+  getGeminiProjects: vi.fn(async () => []),
+}));
```
[the remark this answers is not in the document]
Comment (NiveditJain) [reply __tests__/lib/projects.test.ts:45]: Fixed in a977cd9 (`__tests__/lib/projects.test.ts`). Added three Gemini-specific tests mirroring the existing Pi/Cursor/OpenCode patterns:

1. **Gemini-only inclusion** — when no Claude folder exists, a Gemini project shows up in the merged result with `cli: ["gemini"]`.
2. **Cross-CLI merge by encoded slug** — a Claude folder and a Gemini project sharing `-home-u-shared` collapse into one row with `cli: ["claude", "gemini"]` and the newer `lastModified` (Sept 2026 from Gemini wins over April 2026 from Claude).
3. **Reject-fallback** — when `getGeminiProjects()` rejects, the Claude row still surfaces (other providers don't get blocked by Gemini's failure).

Suite is now 30/30 (was 27).

File: CHANGELOG.md [hunk 3]
```diff
@@ -3,6 +3,7 @@
...
...g builtin policies (block-sudo, block-rm-rf, sanitize-api-keys, …) fire unchanged on Gemini sessions. MCP tool names (`mcp_<server>_<tool>` pattern) and Skills tool names pass through unchanged. The policy evaluator emits Gemini's flat `{decision: "deny", reason}` deny shape (preferred per Gemini's "Golden Rule" exit-0 contract), `{hookSpecificOutput: {hookEventName, additionalContext}}` for context injection on BeforeAgent / AfterTool / SessionStart, and `{decision: "block", reason}` on AfterAgent for force-retry semantics matching Claude's exit-2-from-Stop "do this before stopping" pattern. Path-protection (`isAgentInternalPath` + `isAgentSettingsFile`) covers `~/.gemini/` and `.gemini/settings.json`. Frontend: `lib/cli-registry.ts` adds a `Gemini CLI` entry with a sky-blue badge; `lib/projects.ts` merges Gemini projects into `/projects`; `app/project/[name]` and `/session/[id]` extend the external-CLI fallback chain. Also ships this repo's own `.gemini/settings.json` so contributors using `gemini` get hooks active automatically — uses `$GEMINI_PROJECT_DIR` for resolver stability (Gemini also sets `$CLAUDE_PROJECT_DIR` as a back-compat alias). Verified against gemini-cli v0.40.1.
```
[the remark this answers is not in the document]
Comment (NiveditJain) [reply CHANGELOG.md]: Fixed in a977cd9 (`CHANGELOG.md`). Both Unreleased entries now end with `(#277)`:

- The Features bullet ("Add Gemini CLI integration (beta) across hooks, activity dashboard, session viewer, and projects listing. … Verified against gemini-cli v0.40.1 (#277).")
- The Docs bullet ("README: add Gemini CLI to the supported-CLIs intro line and visual list … (#277).")

Also added a new Fixes bullet for this CR follow-up commit.

File: lib/gemini-projects.ts [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+  for (const s of sessions) {
+    const basename = s.cwd.split("/").filter(Boolean).pop() ?? s.cwd;
+    const existing = byCwd.get(s.cwd);
+    if (!existing || s.fileMtime.getTime() > existing.mtime.getTime()) {
+      byCwd.set(s.cwd, { mtime: s.fileMtime, basename });
+    }
+  }
+  const folders: ProjectFolder[] = [...byCwd.entries()].map(([cwd, { mtime, basename }]) => ({
+    name: basename,
+    path: cwd,
```
[the remark this answers is not in the document]
Comment (NiveditJain) [reply lib/gemini-projects.ts]: Fixed in a977cd9 (`lib/gemini-projects.ts`). `ProjectFolder.name` now uses `encodeFolderName(cwd)` instead of the lossy basename, exactly as suggested:

```diff
- const basename = s.cwd.split("/").filter(Boolean).pop() ?? s.cwd;
- byCwd.set(s.cwd, { mtime: s.fileMtime, basename });
+ byCwd.set(s.cwd, { mtime: s.fileMtime });
…
- const folders: ProjectFolder[] = [...byCwd.entries()].map(([cwd, { mtime, basename }]) => ({
-   name: basename,
+ const folders: ProjectFolder[] = [...byCwd.entries()].map(([cwd, { mtime }]) => ({
+   name: encodeFolderName(cwd),
```

Cross-CLI `mergeProjectFolders` now unions on the same key as Pi/Cursor/OpenCode (their tests already exercise this — confirmed by the new merge-by-encoded-slug test in `__tests__/lib/projects.test.ts`).

File: lib/gemini-sessions.ts [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+  type QueueOperationEntry,
+  type ContentBlock,
+  type LogSource,
+} from "./log-entries";
+
+// ── Paths ──
+
+const UUID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
+const SESSION_FILE_RE = /^session-(\d{4}-\d{2}-\d{2}T\d{2}-\d{2})-([0-9a-f]{8})\.jsonl$/i;
+
```
[the remark this answers is not in the document]
Comment (NiveditJain) [reply lib/gemini-sessions.ts:47]: Fixed in a977cd9 (`lib/gemini-sessions.ts`). Loosened to accept any timestamp shape between `session-` and the 8-hex suffix:

```diff
- const SESSION_FILE_RE = /^session-(\d{4}-\d{2}-\d{2}T\d{2}-\d{2})-([0-9a-f]{8})\.jsonl$/i;
+ const SESSION_FILE_RE = /^session-(.+)-([0-9a-f]{8})\.jsonl$/i;
```

The first-line `sessionId` validation in `findGeminiTranscript` remains the load-bearing safety check, so the more permissive regex is safe even if Gemini ships filenames with seconds, fractional seconds, or other variations in the future.

File: lib/gemini-sessions.ts [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+    for (const f of files) {
+      const m = SESSION_FILE_RE.exec(f);
+      if (!m || m[2].toLowerCase() !== wantPrefix) continue;
+      const candidate = resolve(chatsDir, f);
+      if (!candidate.startsWith(`${chatsDir}${sep}`)) continue;
+      if (!existsSync(candidate)) continue;
+      // Confirm the full sessionId — the 8-hex prefix isn't unique on its own.
+      try {
+        const head = readFileSync(candidate, "utf-8");
+        const firstLine = head.indexOf("\n") >= 0 ? head.slice(0, head.indexOf("\n")) : head;
```
[the remark this answers is not in the document]
Comment (NiveditJain) [reply lib/gemini-sessions.ts]: Fixed in a977cd9 (`lib/gemini-sessions.ts`). Replaced `readFileSync(candidate, "utf-8")` with a new `readFirstLineSync` helper that uses `openSync` + `readSync` of a 4 KB buffer:

```ts
function readFirstLineSync(path: string): string | null {
  let fd: number;
  try { fd = openSync(path, "r"); } catch { return null; }
  try {
    const buf = Buffer.alloc(4096);
    const bytesRead = readSync(fd, buf, 0, buf.length, 0);
    if (bytesRead === 0) return null;
    const text = buf.subarray(0, bytesRead).toString("utf-8");
    const nl = text.indexOf("\n");
    return nl >= 0 ? text.slice(0, nl) : text;
  } catch { return null; } finally {
    try { closeSync(fd); } catch { /* best-effort */ }
  }
}
```

The metadata header sits well within 4 KB; `getGeminiSessionLog()` re-reads the full file once we've matched the sessionId, so the duplicate-read concern is now bounded to the cheap header check.
